components/resolved_ts/src/resolver.rs

```diff
@@ -98,11 +110,10 @@ impl Resolver {
         // Resolved ts never decrease.
         self.resolved_ts = cmp::max(self.resolved_ts, new_resolved_ts);
 
-        // Update region read progress
-        // TODO: should use `RegionReadProgress::update_safe_ts` to avoid direct
-        // write to `safe_ts`
-        self.read_progress
-            .fetch_max_safe_ts(self.resolved_ts.into_inner());
+        // Publish an `(apply index, safe ts)` item into the region read progress
+        if let Some(rrp) = &self.read_progress {
```

Comment:
How can it prevent safe_ts being updated before apply index?

Reply:
`update_safe_ts` will not update `safe_ts` if `apply_index` is smaller than the passed `tracked_index` and instead putting the `(tracked_index, resolved_ts)` into the `RegionReadProgress::pending_items`, and update the `safe_ts` after `apply index` is updated by `RegionReadProgress::update_applied`

Reply:
What if `resolved_ts` is updated before `tracked_index` is updated? Like `apply_index` is updated in apply worker, but resolved ts is updated before `apply_index` is sent.